### ISSUE TITLE
OpenMMParameterSet - repeats of residues

### ISSUE THREAD

**Issue by rafwiewiora** [ParmEd/ParmEd#629]
Similar thing again - should it be in ParmEd or in my script?

atomic_ions.lib has some identical residue templates with different names - e.g. `K` and `K+` - now these both get written to the ffxml. This won't cause a bug in OpenMM - but it's kinda this 'undocumented behavior' of ForceField so we shouldn't do this - not even quite sure how to do this yet, one way would be too look for identical residue signatures in OpenMM - but that'd be post-writing the ffxml editing perhaps.

**Comment by swails** [comment]
Why is this unallowed in OpenMM -- where does it cause issues?  The residues have different names.  Last time I checked, OpenMM still matched residues based on name (it just didn't match atoms in each residue based on name).  This is done in Amber to support both non-PDB and PDB standard names (that are common).

**Comment by jchodera** [comment]
OpenMM does *not* match residues based on name---it uses a graph matching algorithm to match based on the connectivity of elements. Residues with identical element graphs are not exactly supported---I'm not sure if that currently triggers an exception or not, but the behavior of which residue gets matched in that case is definitely undefined.

**Comment by rafwiewiora** [comment]
Current situation:

* no matching on names - matching is done based on 'residue signature' (which is just a string of elements in the residue and number of occurrences of the elements, e.g. `C5H10`) + identical bonding. 

* if you load in more than one identical residues: they get registered in the templates dictionary under the same template signature key:

 ``` python
def registerResidueTemplate(self, template):
        """Register a new residue template."""
        self._templates[template.name] = template
        signature = _createResidueSignature([atom.element for atom in template.atoms])
        if signature in self._templateSignatures:
            self._templateSignatures[signature].append(template)
        else:
            self._templateSignatures[signature] = [template]
```

 That's where the behavior becomes 'undefined' because you've just introduced an order in that list that depends on the order of loading in these templates. And a bunch of loops in the ForceField `break` when they have found something - all of this `undefined` behavior really comes from the lack of mechanism to check for duplicates, which I guess partially comes from a decision to increase efficiency by not completing loops. 

 So going further, when you come to match your templates with topology:
 ``` python
def _getResidueTemplateMatches(self, res, bondedToAtom):
        template = None
        matches = None
        signature = _createResidueSignature([atom.element for atom in res.atoms()])
        if signature in self._templateSignatures:
            for t in self._templateSignatures[signature]:
                matches = _matchResidue(res, t, bondedToAtom)
                if matches is not None:
                    template = t
                    break
        return [template, matches]
```

You can see you will take the first template that matches on the signature and on bonding (through the `_matchResidue` algorithm - [line 987](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L987)) - so the behavior is somewhat 'defined' - it will always take the first-to-have-been-loaded template that matches. 

Clearly we need to tighten this up for the overriding. I think we're going to need to let go of some efficiency and do these loops over everything - so if more than one matching template found - exception. I wouldn't want to complicate this further by allowing more than one matching template if the templates are identical - OpenMM obviously doesn't have the different Pdb format issue because it doesn't match by name. 

So what is the best way to get rid of these duplicates?

**Comment by rafwiewiora** [comment]
Closing this - solved in #639.
